tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -35,6 +35,104 @@ namespace ROOT {
 unsigned int TTreeProcessorMT::fgMaxTasksPerFilePerWorker = 24U;
 
 namespace Internal {
+
+/// A cluster of entries
+struct EntryCluster {
+   Long64_t start;
+   Long64_t end;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+/// Construct fChain, also adding friends if needed and injecting knowledge of offsets if available.
+void TTreeView::MakeChain(const std::string &treeName, const std::vector<std::string> &fileNames,
+                          const FriendInfo &friendInfo, const std::vector<Long64_t> &nEntries,
+                          const std::vector<std::vector<Long64_t>> &friendEntries)
+{
+   const std::vector<NameAlias> &friendNames = friendInfo.fFriendNames;
+   const std::vector<std::vector<std::string>> &friendFileNames = friendInfo.fFriendFileNames;
+
+   fChain.reset(new TChain(treeName.c_str()));
+   const auto nFiles = fileNames.size();
+   for (auto i = 0u; i < nFiles; ++i) {
+      fChain->Add(fileNames[i].c_str(), nEntries[i]);
+   }
+   fChain->ResetBit(TObject::kMustCleanup);
+
+   fFriends.clear();
+   const auto nFriends = friendNames.size();
+   for (auto i = 0u; i < nFriends; ++i) {
+      const auto &friendName = friendNames[i];
+      const auto &name = friendName.first;
+      const auto &alias = friendName.second;
+
+      // Build a friend chain
+      auto frChain = std::make_unique<TChain>(name.c_str());
+      const auto nFileNames = friendFileNames[i].size();
+      for (auto j = 0u; j < nFileNames; ++j)
+         frChain->Add(friendFileNames[i][j].c_str(), friendEntries[i][j]);
+
+      // Make it friends with the main chain
+      fChain->AddFriend(frChain.get(), alias.c_str());
+      fFriends.emplace_back(std::move(frChain));
+   }
+}
+
+TTreeView::TreeReaderEntryListPair
+TTreeView::MakeReaderWithEntryList(TEntryList &globalList, Long64_t start, Long64_t end)
+{
+   // TEntryList and SetEntriesRange do not work together (the former has precedence).
+   // We need to construct a TEntryList that contains only those entry numbers in our desired range.
+   auto localList = std::make_unique<TEntryList>();
+   // We call GetEntry twice: workaround for ROOT-10113 (the return value for 1 call only could be -1)
+   globalList.GetEntry(0);
+   Long64_t entry = globalList.GetEntry(0);
+   do {
+      if (entry >= end) {
+         break;
+      } else if (entry >= start) {
+         localList->Enter(entry);
+      }
+   } while ((entry = globalList.Next()) >= 0);
+
+   auto reader = std::make_unique<TTreeReader>(fChain.get(), localList.get());
+   return std::make_pair(std::move(reader), std::move(localList));
+}
+
+std::unique_ptr<TTreeReader> TTreeView::MakeReader(Long64_t start, Long64_t end)
+{
+   auto reader = std::make_unique<TTreeReader>(fChain.get());
+   reader->SetEntriesRange(start, end);
+   return reader;
+}
+
+TTreeView::TTreeView() {}
+
+// no-op, we don't want to copy the local TChains
+TTreeView::TTreeView(const TTreeView &) {}
```

Comment:
could be inlined (inspecting a header, if I see a copy-ctor defined out of line, I expect it to be complex)